### PR TITLE
http/endpoint: handle URL-decode errors

### PR DESCRIPTION
**Pull request by alxndrsn**
Previously, non-url-decodable paths would return 500.

### CODE REVIEW

**Comment by alxndrsn** [comment]
I think it would also be reasonable to return 400 (Bad Request).

File: lib/http/endpoint.js [hunk 1]
```diff
@@ -223,6 +223,10 @@ const defaultResultWriter = (result, request, response, next) => {
 // error thrown upstream that is of our own internal format, this handler does
 // the necessary work to translate that error into an HTTP error and send it out.
 const defaultErrorWriter = (error, request, response) => {
+  if (error instanceof URIError) {
```
Comment (matthew-white) [inline lib/http/endpoint.js:226]: I like the idea of returning a proper Problem if a `URIError` is thrown for a badly encoded URL. However, if a `URIError` is thrown for another reason, I wouldn't want to swallow it here. What do you think about confirming here that the URL was badly encoded?

```suggestion
  if (error instanceof URIError && urlDecode(request.url).isEmpty()) {
```

File: lib/http/endpoint.js [hunk 2]
```diff
@@ -223,6 +223,10 @@ const defaultResultWriter = (result, request, response, next) => {
...
+    error = Problem.user.notFound(); // eslint-disable-line no-param-reassign
```
Comment (matthew-white) [inline lib/http/endpoint.js:227]: What do you think of this alternative to reassigning `error`?

```suggestion
    return defaultErrorWriter(Problem.user.notFound(), request, response);
```

File: lib/http/endpoint.js (hunk 1 above)
Comment (alxndrsn) [reply lib/http/endpoint.js:226]: I think this is a good idea, but I wish there was a more explicit way to check if the error came from express's internals.

Comment (alxndrsn) [reply lib/http/endpoint.js:226]: Some properties of the error, which we could check:

```js
> e.status === 400
> e.statusCode === 400
> e.message.match(/^Failed to decode param '.*'$/)
```

See: https://github.com/expressjs/express/blob/4.x/lib/router/layer.js#L171-L177

Comment (alxndrsn) [reply lib/http/endpoint.js]: I've added checks and comments, which may address the concerns.

**Review by matthew-white — APPROVED** [review]

File: Makefile
[the remark this answers is not in the document]
Comment (alxndrsn) [reply Makefile:1]: I've updated the PR against `master`, and this has gone away.

**Pull request merged**